README
README file for CGI::TabPane.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip CGI-TabPane-1.04.tgz
shell>tar mxvf CGI-TabPane-1.04.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run TabPane.pm through you favourite pod2html translator.

If you are using my fancy-pom2.pl, with its 'default.css' file installed in
/apache2/htdocs/css/, you'd do:

shell>perl fancy-pom2.pl html -css TabPane.pm > /apache2/htdocs/TabPane.html

or perhaps something like:

shell>perl fancy-pom2.pl html -css TabPane.pm > /perl/html/site/lib/CGI/TabPane.html

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip CGI-TabPane-1.04.zip
shell>ppm install --location=. CGI-TabPane
shell>del CGI-TabPane-1.04.ppd
shell>del PPM-CGI-TabPane-1.04.tar.gz


CHANGES
Revision history for Perl extension CGI::TabPane.

1.08  Sun Feb 21 12:54:47 2010
	- Remove text 'All rights reserved' (for Debian licensing).
	- Remove POD heads 'Required Modules' and 'Changes'.

1.07  Wed Feb 10 14:01:33 2010
	- MANIFEST.SKIP updated to exclude MYMETA.yml. Add MYMETA.yml.

1.06  Fri Nov 13 13:20:00 2009
	- Run dos2unix
	- Rename Changes.txt to CHANGES

1.05  Sun Jan  9 16:44:00 2005
	- Add a check so we don't access and undef array reference

1.04  Mon Jul 19 14:43:00 2004
	- Change Makefile.PL to coexist with Module::Build
	- Move a few CSS declarations around between the *.css files, to help clean these files up.
			You should notice nothing
	- Change the default margin and padding for forms, from 0 to 5 pixels. See final.css lines 14 .. 15.
		In this case, you should now notice a coloured frame around your forms
	- Note: Firefox V 0.9.1 fails to properly handle the CSS required to display multiple panes stacked
		vertically when the last pane is wider than the panes above it. However it does better job than
		IE V 6.0.2800 at handling the width of the tab on a pane with one tab below a pane of several tabs.
		You are encouraged to report all bugs to the appropriate developers
	- Add t/pod.t to test all PODs

1.03  Tue Feb 03 12:37:29 2004
	- Add warning to docs that the CSS file given by the final_css parameter to new()
		must follow, in the <head> of the HTML, the CSS file given by the style_css
		parameter. Add corresponding comment to examples/test-cgi-tabpane.cgi
	- Only add the infix_html between panes, not also after the last pane as I was doing
	- Cut all unused CSS selectors from *.css. This makes it much easier to combine this
		module with other CSS files
	- Change docs to recommend separating panes vertically by 1 paragraph rather than 2.
		Since the CSS selector 'p' was cut from *.css, your demos will change appearence
		whether you use 1 or 2 paragraphs in this way

1.02  Thu Jan 29 12:37:29 2004
	- Add parameter to new() called current_tab, to allow the user to specify which tab is to be the current tab
		when the first pane is displayed. See POD for details
	- Add parameter to new() called use_cookie, to allow the user to specify whether or not a cookie is used by the
		JavaScript to save the state of which tab is the current tab on the first pane. This option was suggested by
		Enrico Sorcinelli. See POD for details
	- Document limitations: 66 tabs per pane and 99 non-nested panes
	- Change call to new() in examples/test-cgi-tabpane.cgi to include these values:
		current_tab => 'Privacy',
		use_cookie  => 0
	- Fix typo in examples/test-cgi-tabpane.cgi. The parameter to new() was:
			style     => '/css/tabpane/webfx.css',
		and is now:
			style_css => '/css/tabpane/webfx.css',
	- Fix Makefile.PL to exclude DBI, because Makefile.PL was copied carelessly from another project, and to include
		Test::More, because the latter module is used in t/test.t

1.01  Mon Jan 25 12:37:29 2004
	- Fix class name in t/test.t. I copied the test from another module and did not edit it properly
	- Fix typos in luna.css: background-repaet is now background-repeat
	- Fix typo in this file: Tue (below) is now Sun

1.00  Sun Jan 25 12:37:29 2004
	- Original version


js/tabpane/tabpane.js
/*----------------------------------------------------------------------------\
|                               Tab Pane 1.02                                 |
|-----------------------------------------------------------------------------|
|                         Created by Erik Arvidsson                           |
|                  (http://webfx.eae.net/contact.html#erik)                   |
|                      For WebFX (http://webfx.eae.net/)                      |
|-----------------------------------------------------------------------------|
|                  Copyright (c) 1998 - 2003 Erik Arvidsson                   |
|-----------------------------------------------------------------------------|
| This software is provided "as is", without warranty of any kind, express or |
| implied, including  but not limited  to the warranties of  merchantability, |
| fitness for a particular purpose and noninfringement. In no event shall the |
| authors or  copyright  holders be  liable for any claim,  damages or  other |
| liability, whether  in an  action of  contract, tort  or otherwise, arising |
| from,  out of  or in  connection with  the software or  the  use  or  other |
| dealings in the software.                                                   |
| - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - |
| This  software is  available under the  three different licenses  mentioned |
| below.  To use this software you must chose, and qualify, for one of those. |
| - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - |
| The WebFX Non-Commercial License          http://webfx.eae.net/license.html |
| Permits  anyone the right to use the  software in a  non-commercial context |
| free of charge.                                                             |
| - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - |
| The WebFX Commercial license           http://webfx.eae.net/commercial.html |
| Permits the  license holder the right to use  the software in a  commercial |
| context. Such license must be specifically obtained, however it's valid for |
| any number of  implementations of the licensed software.                    |
| - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - |
| GPL - The GNU General Public License    http://www.gnu.org/licenses/gpl.txt |
| Permits anyone the right to use and modify the software without limitations |
| as long as proper  credits are given  and the original  and modified source |
| code are included. Requires  that the final product, software derivate from |
| the original  source or any  software  utilizing a GPL  component, such  as |
| this, is also licensed under the GPL license.                               |
|-----------------------------------------------------------------------------|
| 2002-01-?? | First working version                                          |
| 2002-02-17 | Cleaned up for 1.0 public version                              |
| 2003-02-18 | Changed from javascript uri for anchors to return false        |
| 2003-03-03 | Added dispose methods to release IE memory                     |
|-----------------------------------------------------------------------------|
| Dependencies: *.css           a css file to define the layout               |
|-----------------------------------------------------------------------------|
| Created 2002-01-?? | All changes are in the log above. | Updated 2003-03-03 |
\----------------------------------------------------------------------------*/

// This function is used to define if the browser supports the needed
// features
function hasSupport() {

	if (typeof hasSupport.support != "undefined")
		return hasSupport.support;
	
	var ie55 = /msie 5\.[56789]/i.test( navigator.userAgent );
	
	hasSupport.support = ( typeof document.implementation != "undefined" &&
			document.implementation.hasFeature( "html", "1.0" ) || ie55 )
			
	// IE55 has a serious DOM1 bug... Patch it!
	if ( ie55 ) {
		document._getElementsByTagName = document.getElementsByTagName;
		document.getElementsByTagName = function ( sTagName ) {
			if ( sTagName == "*" )
				return document.all;
			else
				return document._getElementsByTagName( sTagName );
		};
	}

	return hasSupport.support;
}

///////////////////////////////////////////////////////////////////////////////////
// The constructor for tab panes
//
// el : HTMLElement		The html element used to represent the tab pane
// bUseCookie : Boolean	Optional. Default is true. Used to determine whether to us
//						persistance using cookies or not
//
function WebFXTabPane( el, bUseCookie ) {
	if ( !hasSupport() || el == null ) return;
	
	this.element = el;
	this.element.tabPane = this;
	this.pages = [];
	this.selectedIndex = null;
	this.useCookie = bUseCookie != null ? bUseCookie : true;
	
	// add class name tag to class name
	this.element.className = this.classNameTag + " " + this.element.className;
	
	// add tab row
	this.tabRow = document.createElement( "div" );
	this.tabRow.className = "tab-row";
	el.insertBefore( this.tabRow, el.firstChild );

	var tabIndex = 0;
	if ( this.useCookie ) {
		tabIndex = Number( WebFXTabPane.getCookie( "webfxtab_" + this.element.id ) );
		if ( isNaN( tabIndex ) )
			tabIndex = 0;
	}
	this.selectedIndex = tabIndex;
	
	// loop through child nodes and add them
	var cs = el.childNodes;
	var n;
	for (var i = 0; i < cs.length; i++) {
		if (cs[i].nodeType == 1 && cs[i].className == "tab-page") {
			this.addTabPage( cs[i] );
		}
	}
}

WebFXTabPane.prototype.classNameTag = "dynamic-tab-pane-control";

WebFXTabPane.prototype.setSelectedIndex = function ( n ) {
	if (this.selectedIndex != n) {
		if (this.selectedIndex != null && this.pages[ this.selectedIndex ] != null )
			this.pages[ this.selectedIndex ].hide();
		this.selectedIndex = n;
		this.pages[ this.selectedIndex ].show();
		
		if ( this.useCookie )
			WebFXTabPane.setCookie( "webfxtab_" + this.element.id, n );	// session cookie
	}
};
	
WebFXTabPane.prototype.getSelectedIndex = function () {
	return this.selectedIndex;
};
	
WebFXTabPane.prototype.addTabPage = function ( oElement ) {
	if ( !hasSupport() ) return;
	
	if ( oElement.tabPage == this )	// already added
		return oElement.tabPage;

	var n = this.pages.length;
	var tp = this.pages[n] = new WebFXTabPage( oElement, this, n );
	tp.tabPane = this;
	
	// move the tab out of the box
	this.tabRow.appendChild( tp.tab );
			
	if ( n == this.selectedIndex )
		tp.show();
	else
		tp.hide();
		
	return tp;
};
	
WebFXTabPane.prototype.dispose = function () {
	this.element.tabPane = null;
	this.element = null;		
	this.tabRow = null;
	
	for (var i = 0; i < this.pages.length; i++) {
		this.pages[i].dispose();
		this.pages[i] = null;
	}
	this.pages = null;
};



// Cookie handling
WebFXTabPane.setCookie = function ( sName, sValue, nDays ) {
	var expires = "";
	if ( nDays ) {
		var d = new Date();
		d.setTime( d.getTime() + nDays * 24 * 60 * 60 * 1000 );
		expires = "; expires=" + d.toGMTString();
	}

	document.cookie = sName + "=" + sValue + expires + "; path=/";
};

WebFXTabPane.getCookie = function (sName) {
	var re = new RegExp( "(\;|^)[^;]*(" + sName + ")\=([^;]*)(;|$)" );
	var res = re.exec( document.cookie );
	return res != null ? res[3] : null;
};

WebFXTabPane.removeCookie = function ( name ) {
	setCookie( name, "", -1 );
};








///////////////////////////////////////////////////////////////////////////////////
// The constructor for tab pages. This one should not be used.
// Use WebFXTabPage.addTabPage instead
//
// el : HTMLElement			The html element used to represent the tab pane
// tabPane : WebFXTabPane	The parent tab pane
// nindex :	Number			The index of the page in the parent pane page array
//
function WebFXTabPage( el, tabPane, nIndex ) {
	if ( !hasSupport() || el == null ) return;
	
	this.element = el;
	this.element.tabPage = this;
	this.index = nIndex;
	
	var cs = el.childNodes;
	for (var i = 0; i < cs.length; i++) {
		if (cs[i].nodeType == 1 && cs[i].className == "tab") {
			this.tab = cs[i];
			break;
		}
	}
	
	// insert a tag around content to support keyboard navigation
	
	
	var a = document.createElement( "A" );
	this.aElement = a;
	a.href = "#";
	a.onclick = function () { return false; };
	while ( this.tab.hasChildNodes() )
		a.appendChild( this.tab.firstChild );
	this.tab.appendChild( a );

	
	// hook up events, using DOM0
	var oThis = this;
	this.tab.onclick = function () { oThis.select(); };
	this.tab.onmouseover = function () { WebFXTabPage.tabOver( oThis ); };
	this.tab.onmouseout = function () { WebFXTabPage.tabOut( oThis ); };
}

WebFXTabPage.prototype.show = function () {
	var el = this.tab;
	var s = el.className + " selected";
	s = s.replace(/ +/g, " ");
	el.className = s;
	
	this.element.style.display = "block";
};

WebFXTabPage.prototype.hide = function () {
	var el = this.tab;
	var s = el.className;
	s = s.replace(/ selected/g, "");
	el.className = s;

	this.element.style.display = "none";
};
	
WebFXTabPage.prototype.select = function () {
	this.tabPane.setSelectedIndex( this.index );
};
	
WebFXTabPage.prototype.dispose = function () {
	this.aElement.onclick = null;
	this.aElement = null;
	this.element.tabPage = null;
	this.tab.onclick = null;
	this.tab.onmouseover = null;
	this.tab.onmouseout = null;
	this.tab = null;
	this.tabPane = null;
	this.element = null;
};

WebFXTabPage.tabOver = function ( tabpage ) {
	var el = tabpage.tab;
	var s = el.className + " hover";
	s = s.replace(/ +/g, " ");
	el.className = s;
};

WebFXTabPage.tabOut = function ( tabpage ) {
	var el = tabpage.tab;
	var s = el.className;
	s = s.replace(/ hover/g, "");
	el.className = s;
};


// This function initializes all uninitialized tab panes and tab pages
function setupAllTabs() {
	if ( !hasSupport() ) return;

	var all = document.getElementsByTagName( "*" );
	var l = all.length;
	var tabPaneRe = /tab\-pane/;
	var tabPageRe = /tab\-page/;
	var cn, el;
	var parentTabPane;
	
	for ( var i = 0; i < l; i++ ) {
		el = all[i]
		cn = el.className;

		// no className
		if ( cn == "" ) continue;
		
		// uninitiated tab pane
		if ( tabPaneRe.test( cn ) && !el.tabPane )
			new WebFXTabPane( el );
	
		// unitiated tab page wit a valid tab pane parent
		else if ( tabPageRe.test( cn ) && !el.tabPage &&
					tabPaneRe.test( el.parentNode.className ) ) {
			el.parentNode.tabPane.addTabPage( el );			
		}
	}
}

function disposeAllTabs() {
	if ( !hasSupport() ) return;
	
	var all = document.getElementsByTagName( "*" );
	var l = all.length;
	var tabPaneRe = /tab\-pane/;
	var cn, el;
	var tabPanes = [];
	
	for ( var i = 0; i < l; i++ ) {
		el = all[i]
		cn = el.className;

		// no className
		if ( cn == "" ) continue;
		
		// tab pane
		if ( tabPaneRe.test( cn ) && el.tabPane )
			tabPanes[tabPanes.length] = el.tabPane;
	}
	
	for (var i = tabPanes.length - 1; i >= 0; i--) {
		tabPanes[i].dispose();
		tabPanes[i] = null;
	}
}


// initialization hook up

// DOM2
if ( typeof window.addEventListener != "undefined" )
	window.addEventListener( "load", setupAllTabs, false );

// IE 
else if ( typeof window.attachEvent != "undefined" ) {
	window.attachEvent( "onload", setupAllTabs );
	window.attachEvent( "onunload", disposeAllTabs );
}

else {
	if ( window.onload != null ) {
		var oldOnload = window.onload;
		window.onload = function ( e ) {
			oldOnload( e );
			setupAllTabs();
		};
	}
	else 
		window.onload = setupAllTabs;
}


lib/CGI/TabPane.pm
package CGI::TabPane;

# Name:
#	CGI::TabPane.
#
# Documentation:
#	POD-style documentation is at the end. Extract it with pod2html.*.
#
# Reference:
#	Object Oriented Perl
#	Damian Conway
#	Manning
#	1-884777-79-1
#	P 114
#
# Note:
#	o Tab = 4 spaces || die.
#
# Author:
#	Ron Savage <ron@savage.net.au>
#	Home page: http://savage.net.au/index.html
#
# Licence:
#	Australian copyright (c) 2004 Ron Savage.
#
#	All Programs of mine are 'OSI Certified Open Source Software';
#	you can redistribute them and/or modify them under the terms of
#	The Artistic License, a copy of which is available at:
#	http://www.opensource.org/licenses/index.html

use strict;
use warnings;
no warnings 'redefine';

use Carp;

require 5.005_62;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use CGI::TabPane ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(

) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);
our $VERSION = '1.08';

# -----------------------------------------------

# Preloaded methods go here.

# -----------------------------------------------

# Encapsulated class data.

{
	my(%_attr_data) =
	(	# Alphabetical order.
		_current_tab	=> '',
		_data			=> '',
		_final_css		=> '/css/tabpane/final.css',
		_html			=> '',
		_infix_html		=> '',
		_prefix_html	=> '',
		_style_css		=> '/css/tabpane/luna.css',
		_suffix_html	=> '',
		_tabpane_js		=> '/js/tabpane/tabpane.js',
		_use_cookie		=> 1,
		_webfxlayout_js	=> '/js/tabpane/webfxlayout.js',
	);

	sub _build
	{
		my($self, $pane_id, $pane, $nested) = @_;
		my($html)	= qq|<div class = "tab-pane" id = "tabPane$pane_id">\n|;
		$html		.= qq|<script type = "text/javascript">tp1 = new WebFXTabPane(document.getElementById("tabPane$pane_id"), $$self{'_use_cookie'});</script>\n| if ($pane_id == 0);

		my($tab_index);

		for my $tab (0 .. $#$pane)
		{
			# Arbitrarily limit the number of tabs per pane to 66, even though I hate to do this.
			# This $page_id is the HTML id (name) of the tab within the page.

			my($page_id)	= 66 * $nested + $tab;
			my(@key)		= keys %{$$pane[$tab]};
			$tab_index		= $tab if (! $nested && $key[0] && ($key[0] eq $$self{'_current_tab'}) );
			$html			.= ($pane_id == 0) ? qq|<div class = "tab-page" id = "tabPage$page_id">\n| : qq|<div class = "tab-page">\n|;
			$html			.= qq|<h2 class = "tab">$key[0]</h2>\n|;
			$html			.= qq|<script type = "text/javascript">tp1.addTabPage(document.getElementById("tabPage$page_id") );</script>\n| if ($pane_id == 0);

			if (ref($$pane[$tab]{$key[0]}) eq 'HASH')
			{
				my(@legend)	= keys %{$$pane[$tab]{$key[0]} };
				$html		.=	qq|<fieldset>\n| .
								qq|<legend>$legend[0]</legend>\n| .
								qq|$$pane[$tab]{$key[0]}{$legend[0]}\n| .
								qq|</fieldset>|;
			}
			elsif (ref($$pane[$tab]{$key[0]}) eq 'ARRAY')
			{
				# Arbitrarity limit the number of non-nested panes to 100, even though I hate to do this.
				# This (99 + $pane_id) is the id (name) of the first nested pane.

				$html .= $self -> _build( (99 + $pane_id), $$pane[$tab]{$key[0]}, 1);
			}
			else
			{
				$html .= qq|$$pane[$tab]{$key[0]}\n|;
			}

			$html .= qq|</div>\n|;
		}

		$html .= qq|<script type = "text/javascript">tp1.setSelectedIndex($tab_index);</script>\n| if (! $nested && $tab_index);
		$html .= qq|</div>\n|;

	}	# End of _build.

	sub _default_for
	{
		my($self, $attr_name) = @_;

		$_attr_data{$attr_name};
	}

	sub _standard_keys
	{
		keys %_attr_data;
	}

}	# End of encapsulated class data.

# -----------------------------------------------

sub get
{
	my($self, $arg) = @_;

	my($result);

	for my $attr_name ($self -> _standard_keys() )
	{
		my($arg_name) = $attr_name =~ /^_(.*)/;

		if (exists($$self{$attr_name}) && ($arg eq $arg_name) )
		{
			$result = $$self{$attr_name};
		}
	}

	$result;

}	# End of get.

# -----------------------------------------------

sub new
{
	my($caller, %arg)	= @_;
	my($caller_is_obj)	= ref($caller);
	my($class)			= $caller_is_obj || $caller;
	my($self)			= bless({}, $class);

	for my $attr_name ($self -> _standard_keys() )
	{
		my($arg_name) = $attr_name =~ /^_(.*)/;

		if (exists($arg{$arg_name}) )
		{
			$$self{$attr_name} = $arg{$arg_name};
		}
		elsif ($caller_is_obj)
		{
			$$self{$attr_name} = $$caller{$attr_name};
		}
		else
		{
			$$self{$attr_name} = $self -> _default_for($attr_name);
		}
	}

	croak(__PACKAGE__ . ". You must supply a value for the 'data' parameter") if (! $$self{'_data'});

	$$self{'_html'} .= $$self{'_prefix_html'};

	for my $p (0 .. $#{$$self{'_data'} })
	{
		$$self{'_html'} .= $self -> _build($p, $$self{'_data'}[$p], 0);
		$$self{'_html'} .= $$self{'_infix_html'} if ($p < $#{$$self{'_data'} });
	}

	$$self{'_html'} .= $$self{'_suffix_html'};

	return $self;

}	# End of new.

# -----------------------------------------------

sub set
{
	my($self, %arg) = @_;

	for my $attr_name ($self -> _standard_keys() )
	{
		my($arg_name) = $attr_name =~ /^_(.*)/;

		if (exists($arg{$arg_name}) )
		{
			$$self{$attr_name} = $arg{$arg_name};
		}
	}

}	# End of set.

# -----------------------------------------------

1;

__END__

=head1 NAME

C<CGI::TabPane> - Support panes with clickable tabs

=head1 Synopsis

	use CGI::TabPane;

	CGI::TabPane -> new(data => [...]) -> get('html');

=head1 Description

C<CGI::TabPane> is a pure Perl module.

It is a wrapper around the superb JavaScript package 'Tab Pane' by Erik Arvidsson.

Erik's article on Tab Pane is here: http://webfx.eae.net/dhtml/tabpane/tabpane.html

I have simplified some of Erik's files, and renamed some, to make shipping and installing easier.

=head1 Installation

The makefile will install /perl/site/lib/CGI/TabPane.pm.

You must manually install <Document Root>/css/tabpane/*.[css|png] and <Document Root>/js/tabpane/*.js.

If you choose to put the CSS elsewhere, you'll need to call new(final_css => '/new/path/final.css').
Similarly for style_css.

Note: If you put webfxlayout.css elsewhere, you'll have to edit webfxlayout.js.

If you choose to put the JavaScript elsewhere, you'll need to call new(tabpane_js => '/new/path/tabpane.css').
Similarly for webfxlayout_js.

These options can be used together, and can be passed into C<set()> rather than C<new()>.

=head1 Distributions

This module is available both as a Unix-style distro (*.tgz) and an
ActiveState-style distro (*.ppd). The latter is shipped in a *.zip file.

See http://savage.net.au/Perl-modules.html for details.

See http://savage.net.au/Perl-modules/html/installing-a-module.html for
help on unpacking and installing each type of distro.

=head1 Constructor and initialization

new(...) returns a C<CGI::TabPane> object.

This is the class's contructor.

Usage: CGI::TabPane -> new().

This method takes a set of parameters. Only the data parameter is mandatory.

For each parameter you wish to use, call new as new(param_1 => value_1, ...).

=over 4

=item current_tab

This value is a string, the name of tab, used to specify which tab on the first pane is to be the current tab
when that pane is displayed.

For example, tab 'Search' could contain a CGI form, with that tab being the default or current tab the first time the
CGI script is called to generate output. Then, when the submit button on that form is clicked, the CGI script could
generate the next lot of output with a different tab, e.g. tab 'Results', being the current tab.

There is no provision for controlling which tab is the current tab on any pane after the first pane.

The default value is ''.

This parameter is optional.

Note: In the JavaScript, the tabs are numbered left to right, starting at 0.

=item data

This value holds all the information required to build the panes and tabs per pane.

See the sections below called 'Terminology' and 'The Structure of the Data' for details.

The default value is '' (the empty string), which will die because you must pass in an array ref.

This parameter is mandatory.

=item final_css

This value is the name of a CSS file. The module ships with a suitable file called final.css.

Warning: This file must be output in the <head> of your HTML page C<after> the CSS file you
choose with the style_css parameter. See examples/test-cgi-tabpane.cgi for a demo.

Warning: Do not edit final.css. Even tiny changes will have horrible effects on the output.

The default value is '/css/tabpane/final.css'.

This parameter is optional.

=item html

Normally you would not do this, but if you want, you can pass in a value for the C<html> parameter, and the
HTML generated by this module will be appended to your initial value.

The default value is ''.

This parameter is optional.

=item infix_html

This value is a string of HTML to be inserted between panes, to separate them vertically.

A typical value might be: '<p>&nbsp;</p>'.

The default value is '' (the empty string).

This parameter is optional.

=item prefix_html

This value is a string of HTML to be inserted before the first pane.

A typical value might be: '<p>&nbsp;</p>'.

The default value is '' (the empty string).

This parameter is optional.

=item style_css

This value is the name of a CSS file which determines the overall style of the output. The module ships with
three suitable files:

=over 4

=item luna.css

=item webfx.css

=item winclassic.css

=back

The default value is '/css/tabpane/luna.css'.

This parameter is optional.

=item suffix_html

This value is a string of HTML to be inserted after the last pane.

A typical value might be: '<p>&nbsp;</p>'.

The default value is '' (the empty string).

This parameter is optional.

=item tabpane_js

This value is the name of a JavaScript file. The module ships with a suitable file called tabpane.js.

Warning: Do not edit tabpane.js. Even tiny changes will have horrible effects on the output.

The default value is '/js/tabpane/tabpane.js'.

This parameter is optional.

=item use_cookie

This value is a Boolean, 0 or 1, which respectively deactivates or activates the persistance option of the WebFXTabPane
class in tabpane.js. Activation means the JavaScript uses a cookie to save the state of which tab on the first pane is
the current tab.

The default value is 1, since that's the default in the WebFXTabPane class, and so that's what will have been used in
the past when this module was run without this parameter.

There is no provision for saving the state of which tab is the current tab on any pane after the first pane, because
the name of the cookie does not include any indicator as to which pane the current tab belongs to.

This parameter is optional.

Historical note: The word Boolean must have a capital B because it's based on the name of a person - George Boole.

=item webfxlayout_js

This value is the name of a JavaScript file. The module ships with a suitable file called webfxlayout.js.

Warning: Do not edit webfxlayout.js. Even tiny changes will have horrible effects on the output.

The default value is '/js/tabpane/webfxlayout.js'.

This parameter is optional.

=back

=head1 Terminology

The comments below this diagram define a few terms.

The example shipped with this module will make things much clearer. See the examples/ directory.

Alternately, go straight to Erik's demo at: http://webfx.eae.net/dhtml/tabpane/demo.html

	[Tab 1-1] [Tab 1-2] [TAB 1-3] [Tab 1-4]
	[===============================================]
	[ Text for tab 1-3                              ]
	[===============================================]

	[TAB 2-1] [Tab 2-2]
	[==============================]
	[ +Legend--------------------+ ]
	[ |Text for tab 2-1          | ]
	[ +--------------------------+ ]
	[==============================]

=over 4

=item Pane

We have 2 'panes', one on top of the other.

Panes are tiled vertically.

Use the C<infix_html> parameter to C<new()> or C<set()> to change the vertical separation of the panes.

=item Tab

The top pane has 4 'tabs' side-by-side. The bottom pane has 2 tabs.

Tabs are tiled horizontally.

=item Current tab

The current tab within a pane is the tab most-recently clicked.

Tab 3 in the 1st pane and tab 1 in the 2nd pane are in upper case in the diagram to indicate they are the
'current tab' in each pane. The upper case above is just something I made up for the purposes of writing this
document. Of course, text is not converted to upper case by this module.

	Note: Tabs can be clicked with a mouse, and they can be clicked by executing code.

=item Current text

The text associated with the current tab is called the 'current text'.

=item Body

When a tab is clicked it becomes the current tab, and its text becomes the current text, and this text is displayed
in the 'body' of the pane, completely replacing any existing text being displayed in the body.

=item Legend

It is possible to use the HTML tag 'legend' to cause a box to be drawn by the web client (browser) around the current
text.

Further, it is possible to get the web client to write a string of text on top of the upper left part of this box.

Such a string is called the 'legend'. See above for an example. See the examples/ directory for a better example.

=back

=head1 The Structure of the Data

Our first problem is to define a data structure which allows us to neatly supply our own data to populate a set
of panes and each set of tabs per pane.

=over 4

=item Data for a set of panes

The data structure defining the panes is an array ref. Each element of the array provides data for 1 pane.

So, the above diagram will be something like:

	[<Data for 1st pane>, <Data for 2nd pane>].

=item Data for a set of tabs

The data structure defining the tabs for 1 pane is an array ref. Each element of the array provides data for 1 tab.

So, the above diagram will be something like:

	<Data for 1st pane>:

	[<Data for 1st tab of 1st pane>, <Data for 2nd tab of 1st pane>, ...].

	<Data for 2nd pane>:

	[<Data for 1st tab of 2nd pane>, <Data for 2nd tab of 2nd pane>].

=item Data for 1 tab without a legend

In the absence of a legend, the data for 1 tab is a hash ref:

	<Data for any tab (without a legend) of any pane>:

	{'Text for tab' => 'Text for body of pane'}

Recall from the discussion of 'body' above: 'Text for body of pane' means the text which is to be displayed when
this particular tab if clicked.

In the diagram, then, the data for the 3rd tab of the 1st pane is:

	{'Tab 1-3' => 'Text for tab 1-3'}

=item Data for 1 tab with a legend

In the presence of a legend, the data for 1 tab is a hash ref within a hash ref:

	<Data for any tab (with a legend) of any pane>:

	{'Text for tab' => {'Text for legend' => 'Text for body of pane'} }

In the diagram, then, the data for the 1st tab of the 2nd pane is:

	{'Tab 2-1' => {'Legend' => 'Text for tab 2-1'} }

=item Data for nested panes and tabs

In the presence of nested panes, the data for 1 tab is an array ref within a hash ref:

	<Data for a nested pane>:

	{'Text for tab' => [<Data for 1st nested tab>, <Data for 2nd nested tab>, ...]}

Warning: This module will only handle 2 levels in this hierarchy of nesting, since it uses hard-coded CSS tokens for
the classes of the <div>s involved. By '2 levels' I mean 1 outer pane and 1 inner pane.

Warning: To have one nested pane, with its own set of tabs, makes sense. To add more complexity than that surely
means your design is too complex. By 'more complexity' I mean 2 panes displayed vertically within 1 pane.
So don't do that.

=back

Lastly, the entire data structure for the diagram could be:

	[      # All panes
	    [  # 1st pane
	        {'Tab 1-1' => 'Text for tab 1-1'},
	        {'Tab 1-2' => 'Text for tab 1-2'},
	        {'Tab 1-3' => 'Text for tab 1-3'},
	        {'Tab 1-4' => 'Text for tab 1-4'},
	    ],
	    [  # 2nd pane
	        {'Tab 2-1' => {'Legend' => 'Text for tab 2-1'} },
	        {'Tab 2-2' => 'Text for tab 2-2'},
	    ],
	]

You are strongly urged to examine the demo examples/test-cgi-tabpane.cgi to get an understanding of how to construct
the required data structure.

=head1 Limitations

The numbers of the panes, 0 .. N, and the numbers of tabs per pane, 0 .. N, are used to generate strings which in turn
as used as names of things in the JavaScript and HTML.

The non-Perl code will only work if these names are unique. So, you are limited to 99 panes and 66 tabs per pane.

The Perl code does not check to ensure you are within in these limits.

=head1 Method: get($name_of_thing_to_get)

Returns a string.

The 'name of the thing to get' is any of the parameters which can be passed in to C<new()>:

=over 4

=item data

=item final_css

=item html

This is the one which you absolutely must call.

=item infix_html

=item prefix_html

=item style_css

=item suffix_html

=item tabpane_js

=item webfxlayout_js

=back

In each case, the current value of the parameter held within the C<CGI::TabPane> object is returned.

See the demo examples/test-cgi-tabpane.cgi for an example.

=head1 Example code

See the examples/ directory in the distro.

=head1 Related Modules

CGI::Explorer, also one of my modules, is a pure Perl wrapper around another superb package from the House of EAE,
this time 'XTree' by Emil A Eklund.

Emil and Erik share the web site http://webfx.eae.net/ - please drop in there and express your thanx.

=head1 Author

C<CGI::TabPane> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2004.

Home page: http://savage.net.au/index.html

=head1 Copyright

Australian copyright (c) 2004, Ron Savage.
	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut
